### PR TITLE
[Inductor] [Doc] Add tutorial for Max-autotune Support on CPU as a prototype feature for PyTorch 2.5

### PR DESCRIPTION
**Pull request by chunyuan-w**
## Description
 Max-autotune Support on CPU with GEMM Template is proposed as a prototype feature for PyTorch 2.5. This PR adds the tutorial for this feature.

Rendered version: [link](https://github.com/chunyuan-w/tutorials/blob/chunyuan/max-autotune-pr/prototype_source/max_autotune_on_CPU_tutorial.rst).

### CODE REVIEW

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,198 @@
+Max-autotune Support on CPU with GEMM Template Tutorial
```
Comment (jgong5) [inline prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst:1]: GEMM template is an implementation detail. Suggest to name the title from users perspective: "Use max-autotune compilation on CPU to gain further performance boost". Then in the content of the tutorial, we can talk about perf boost in GEMMs with the template implementation etc.

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst
Comment (jgong5) [inline prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst:1]: There are docs with name prefixed `torch.compiler_` in `docs/source`. Shall we move the doc there instead?

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+------------
+To activate the ``max-autotune`` mode in PyTorch, set the ``mode`` argument to ``max-autotune`` when compiling your model using ``torch.compile``.
+If you prefer to bypass the tuning process and always use the CPP template implementations, you can configure this via an environment variable: 
+``export TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS=CPP``.
+
+
+Example code
+------------
+The below code is an example of using the ``max-autotune`` mode on a simple neural network with a linear layer followed by a ReLU activation.
+You could run the example code by setting this environment variable ``export TORCHINDUCTOR_FREEZING=1``.
```
Comment (jgong5) [inline prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst:29]: Since we only support frozen model with `torch.no_grad` or the inference mode, shall we highlight this fact here?

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,198 @@
+Max-autotune Support on CPU with GEMM Template Tutorial
+==============================================================
+
+**Author**: `Jiong Gong <https://github.com/jgong5>`__, `Leslie Fang <https://github.com/leslie-fang-intel>`__, `Chunyuan Wu <https://github.com/chunyuan-w>`__
+
+Prerequisites:
+----------------
+-  `torch.compile and TorchInductor concepts in PyTorch <https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html>`__
+
+Introduction
```
Comment (jgong5) [inline prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst:10]: Perhaps, add the RFC link in the intro as the reference too?

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    cpp_packed_gemm_0 0.2142 ms 100.0% 
+    _linear_pointwise 0.2441 ms 87.7% 
+
+
+We could check the generated output code by setting ``export TORCH_LOGS="+output_code"``.
+When CPP template is selected, we won't have ``torch.ops.mkldnn._linear_pointwise.default`` (for bfloat16) or ``torch.ops.mkl._mkl_linear.default`` (for float32)
+in the generated code anymore, instead, we'll find kernel based on CPP GEMM template ``cpp_fused__to_copy_relu_1``
+(only part of the code is demonstrated below for simplicity) with the bias and relu epilogues fused inside the CPP GEMM template kernel.
+
+.. code:: python
```
Comment (jgong5) [inline prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst:89]: The code below is generated on CPUs with AMX support. You may consider to add a note here that the generated code differs per CPU archs and implementation specific, subject to change?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst (hunk 3 above)
Comment (chunyuan-w) [reply prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst]: RFC link has been added

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst (hunk 1 above)
Comment (chunyuan-w) [reply prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst]: Renamed to "Use max-autotune compilation on CPU to gain further performance boost"

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst (hunk 2 above)
Comment (chunyuan-w) [reply prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst]: The request on frozen and `no_grad` has been added.

File: prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst (hunk 4 above)
Comment (chunyuan-w) [reply prototype_source/max_autotune_CPU_with_gemm_template_tutorial.rst]: The note has been added.

**Review by jgong5 — APPROVED** [review]

**Comment by chunyuan-w** [comment]
Hi @svekars could you help review this tutorial targeting PyTorch 2.5?

File: prototype_source/max_autotune_on_CPU_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+If you prefer to bypass the tuning process and always use the CPP template implementations, you can configure this via an environment variable: 
+``export TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS=CPP``.
+
+
+Example code
+------------
+The below code is an example of using the ``max-autotune`` mode on a simple neural network with a linear layer followed by a ReLU activation.
+
+We only support frozen model with ``torch.no_grad`` or the inference mode
+Therefore, you need to set the environment variable ``export TORCHINDUCTOR_FREEZING=1``
```
Comment (agunapal) [inline prototype_source/max_autotune_on_CPU_tutorial.rst:34]: Could you please write a line or 2 on what is the advantage of freezing and what is the difference without freezing

Comment (chunyuan-w) [reply prototype_source/max_autotune_on_CPU_tutorial.rst]: I've added more explanations why freezing is required. Could you please check if that is clearer?

Comment (agunapal) [reply prototype_source/max_autotune_on_CPU_tutorial.rst]: looks good. Thanks!

**Comment by chunyuan-w** [comment]
Hi @svekars thanks for the suggestions! I've updated the PR. Could you help take another look?

**Comment by chunyuan-w** [comment]
Hi @svekars may I know if there's any action required on my side for this tutorial targeting PyTorch 2.5?

File: prototype_source/max_autotune_on_CPU_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
+Use max-autotune compilation on CPU to gain additional performance boost
+==============================================================
```
Comment (svekars) [inline prototype_source/max_autotune_on_CPU_tutorial.rst:2]:
```suggestion
================================================================================
```

File: prototype_source/max_autotune_on_CPU_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,214 @@
+Use max-autotune compilation on CPU to gain additional performance boost
```
Comment (svekars) [inline prototype_source/max_autotune_on_CPU_tutorial.rst:1]:
```suggestion
Using Max-Autotune Compilation on CPU for Better Performance
```

File: prototype_source/max_autotune_on_CPU_tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+==============================================================
+
+**Author**: `Jiong Gong <https://github.com/jgong5>`__, `Leslie Fang <https://github.com/leslie-fang-intel>`__, `Chunyuan Wu <https://github.com/chunyuan-w>`__
+In this tutorial, you will learn how to boost your PyTorch models' performance on CPU by
```
Comment (svekars) [inline prototype_source/max_autotune_on_CPU_tutorial.rst:5]:
```suggestion

In this tutorial, you will learn how to boost your PyTorch models' performance on CPU by 
```

**Review by svekars — APPROVED** [review]

File: prototype_source/max_autotune_on_CPU_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+into your code for enhanced computational efficiency. Dive into the use of advanced
+GEMM templates for faster processing and superior runtime performance.
+
+Prerequisites:
+----------------
+-  `torch.compile and TorchInductor concepts in PyTorch <https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html>`__
+
+Introduction
+------------
+``max-autotune`` mode for the Inductor CPU backend in ``torch.compile`` (`RFC link <https://github.com/pytorch/pytorch/issues/125683>`_)
```
Comment (svekars) [inline prototype_source/max_autotune_on_CPU_tutorial.rst:18]:
```suggestion
The ``max-autotune`` mode for the Inductor CPU backend in ``torch.compile`` (`RFC link <https://github.com/pytorch/pytorch/issues/125683>`_)
```
